### PR TITLE
Questions in database

### LINKED ISSUE #679: `base.get_single_question` does not use questions defined in database
See also #678 .

### PR DESCRIPTION
**Pull request by albertas-jn**
Implements questions in a database.
Don't forget to backup your database before testing!

Closes #679

### CODE REVIEW

File: backend/experiment/static/questionseries_admin.js
Comment (BeritJanssen) [inline backend/experiment/static/questionseries_admin.js:1]: Can you decrease indentation level? 5 spaces (?) makes it less readable IMO.

File: backend/experiment/static/experiment_admin.js
Comment (BeritJanssen) [inline backend/experiment/static/experiment_admin.js:1]: Please decrease indentation to three spaces.

**Review by BeritJanssen — CHANGES_REQUESTED** [review]
I haven't had the chance to test this yet, but here's my two cents:
- in the interest of not bloating the `experiment` app even further, I'd be strongly in favour of moving the question logic (including the `questions` directory within `experiment`) to its own app, which we may creatively call `question`. I know it's a rather painful change to request (sorry), but I think it's less painful to refactor now than later.
- there is no unit test for the `createquestions` command, or for any of the admin views
- right now (since I've only had a superficial look really) I cannot fully understand why `questionseries_admin` and the views it calls are needed, i.e., why this is not possible in vanilla Django admin. Docstrings would be appreciated, also docstrings explaining what the different question models (Question / QuestioInSeries / QuestionInSeries / QuestionGroup) are for

File: backend/experiment/models.py
Comment (BeritJanssen) [inline backend/experiment/models.py:1]: While the naming of the `Experiment`, `ExperimentCollection` and `ExperimentGroup` is probably just as hard to follow for outsiders, the principles behind this are (in my mind) rather similar, so I think we'd benefit from following the same naming conventions, as much as possible.

File: backend/experiment/models.py [hunk 1]
```diff
@@ -279,7 +273,77 @@ def max_score(self):
...
+        return "("+self.key+") "+ self.question
+
+    class Meta:
+        ordering = ["key"]
+
+
+class QuestionGroup(models.Model):
+
+    key = models.CharField(primary_key=True, max_length=128)
+    questions = models.ManyToManyField(Question)
```
[the remark this answers is not in the document]
Comment (albertas-jn) [reply backend/experiment/models.py:313]: I was hoping that people test it first, so the code would make things much clearer, because I think the interface is quite intuitive. `QuestionGroup` is just a collection of questions for convenience to easier add to the `QuestionSeries`, which belongs to `Experiment` and describes the question sequence in experiment.

File: backend/experiment/models.py
Comment (albertas-jn) [reply backend/experiment/models.py:1]: When I started working on this, there was still `ExperimentSeries`, so I tried to follow that convention. To me, "Series" is clearer because that it implies ordering. But of course it can be renamed. What would you suggest?

File: backend/experiment/models.py [hunk 2]
```diff
@@ -279,7 +273,77 @@ def max_score(self):
...
+        verbose_name_plural = "Question Series"
+
+    def __str__(self):
+        return "QuestionSeries object ({}): {} questions".format(self.id, self.questioninseries_set.count())
+
+
+class QuestionInSeries(models.Model):
+
+    question_series = models.ForeignKey(QuestionSeries, on_delete=models.CASCADE)
+    question = models.ForeignKey(Question, on_delete=models.CASCADE)
```
[the remark this answers is not in the document]
Comment (albertas-jn) [reply backend/experiment/models.py:343]: `OneToOneField` to what?

File: backend/experiment/static/questionseries_admin.js
Comment (albertas-jn) [reply backend/experiment/static/questionseries_admin.js:1]: Ah, it's actually tabs that sneaked in to my .js files. Interestingly, when you "View file" in github, it shows different indentation level than in "Files changed" tab.

File: backend/experiment/static/experiment_admin.js
Comment (albertas-jn) [reply backend/experiment/static/experiment_admin.js:1]: Why 3? All our (now) .jsx files have 4 spaces.

**Comment by Evert-R** [comment]
I tried running this branch but I run into this migration problem:
` Operations to perform:
server-1      |   Apply all migrations: admin, auth, contenttypes, experiment, image, participant, result, section, session, sessions, theme
server-1      | Running migrations:
server-1      |   Applying experiment.0034_add_question_model... OK
server-1      | Traceback (most recent call last):
server-1      |   File "manage.py", line 24, in <module>
server-1      |     main()
server-1      |   File "manage.py", line 20, in main
server-1      |     execute_from_command_line(sys.argv)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
server-1      |     utility.execute()
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
server-1      |     self.fetch_command(subcommand).run_from_argv(self.argv)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
server-1      |     self.execute(*args, **cmd_options)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
server-1      |     output = self.handle(*args, **options)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 89, in wrapped
server-1      |     res = handle_func(*args, **kwargs)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 244, in handle
server-1      |     post_migrate_state = executor.migrate(
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 117, in migrate
server-1      |     state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
server-1      |     state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
server-1      |     state = migration.apply(state, schema_editor)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/migration.py", line 126, in apply
server-1      |     operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
server-1      |     self.code(from_state.apps, schema_editor)
server-1      |   File "/server/experiment/migrations/0035_add_question_model_data.py", line 12, in add_default_question_series
server-1      |     experiment.add_default_question_series()
server-1      |   File "/server/experiment/models.py", line 279, in add_default_question_series
server-1      |     for i,question_series in enumerate(EXPERIMENT_RULES[self.rules]().question_series):
server-1      | AttributeError: 'CongoSameDiff' object has no attribute 'question_series'
server-1      |   Applying experiment.0035_add_question_model_data...
server-1 exited with code 1`

**Comment by Evert-R** [comment]
I tried running this branch but run into these migration problems:
 Operations to perform:
server-1      |   Apply all migrations: admin, auth, contenttypes, experiment, image, participant, result, section, session, sessions, theme
server-1      | Running migrations:
server-1      |   Applying experiment.0034_add_question_model... OK
server-1      | Traceback (most recent call last):
server-1      |   File "manage.py", line 24, in <module>
server-1      |     main()
server-1      |   File "manage.py", line 20, in main
server-1      |     execute_from_command_line(sys.argv)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
server-1      |     utility.execute()
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
server-1      |     self.fetch_command(subcommand).run_from_argv(self.argv)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
server-1      |     self.execute(*args, **cmd_options)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
server-1      |     output = self.handle(*args, **options)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/base.py", line 89, in wrapped
server-1      |     res = handle_func(*args, **kwargs)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/core/management/commands/migrate.py", line 244, in handle
server-1      |     post_migrate_state = executor.migrate(
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 117, in migrate
server-1      |     state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
server-1      |     state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
server-1      |     state = migration.apply(state, schema_editor)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/migration.py", line 126, in apply
server-1      |     operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
server-1      |   File "/usr/local/lib/python3.8/site-packages/django/db/migrations/operations/special.py", line 190, in database_forwards
server-1      |     self.code(from_state.apps, schema_editor)
server-1      |   File "/server/experiment/migrations/0035_add_question_model_data.py", line 12, in add_default_question_series
server-1      |     experiment.add_default_question_series()
server-1      |   File "/server/experiment/models.py", line 279, in add_default_question_series
server-1      |     for i,question_series in enumerate(EXPERIMENT_RULES[self.rules]().question_series):
server-1      | AttributeError: 'CongoSameDiff' object has no attribute 'question_series'
server-1      |   Applying experiment.0035_add_question_model_data...
server-1 exited with code 1

**Comment by albertas-jn** [comment]
@Evert-R Fixed, please try again

File: backend/experiment/static/experiment_admin.js
Comment (BeritJanssen) [reply backend/experiment/static/experiment_admin.js:1]: 4 is fine, indeed!

File: backend/experiment/static/questionseries_admin.js
Comment (BeritJanssen) [reply backend/experiment/static/questionseries_admin.js:1]: Better to set your editor to use spaces for indentation.

File: backend/experiment/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (BeritJanssen) [reply backend/experiment/models.py:315]: Docstrings would still be a good idea.

File: backend/experiment/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (BeritJanssen) [reply backend/experiment/models.py:345]: You're absolutely right, of course!

File: backend/experiment/models.py
Comment (BeritJanssen) [reply backend/experiment/models.py:1]: If you want to call this `QuestionSeries`, that's fine. `ExperimentCollection` will be renamed anyway, so it's not necessary to keep consistency there. But perhaps `Question` -> `GroupedQuestion` -> `QuestionGroup` -> `QuestionSeries`?

File: backend/experiment/static/questionseries_admin.js
Comment (drikusroor) [reply backend/experiment/static/questionseries_admin.js:1]: https://youtu.be/SsoOG6ZeyUI?si=jUvctXKmYammM-Jy

File: backend/experiment/static/experiment_admin.js
Comment (drikusroor) [reply backend/experiment/static/experiment_admin.js:1]: Do we use 3 spaces in python? 🤔

Comment (albertas-jn) [reply backend/experiment/static/experiment_admin.js:1]: No, we use 4

**Comment by albertas-jn** [comment]
The commit messages of the changes made since the last review speak for themselves.

I would just like to comment on the changes in `backend/experiment/migrations/0017_experiment_add_questions_field.py`, one of the older migrations. It is the migration that introduced the question field in Experiment. This field is not relevant anymore. The problem was that it called a function for a default value of question keys, `get_default_question_keys()`. This function is obsolete now, but I could not remove it, because the migration file depended on it. To be able to remove the function, I had to edit the old migration file and set default question keys to empty list. Therefore such seemingly odd change in the old migration file.

File: backend/question/migrations/0001_add_question_model.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                ('editable', models.BooleanField(default=True, editable=False)),
+            ],
+            options={
+                'ordering': ['key'],
+            },
+        ),
+        migrations.CreateModel(
+            name='QuestionInSeries',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
```
Comment (BeritJanssen) [inline backend/question/migrations/0001_add_question_model.py:30]: Why do you explicitly create this field?

File: backend/experiment/management/commands/createquestions.py
Comment (BeritJanssen) [inline backend/experiment/management/commands/createquestions.py:1]: Is there a situation where you would want to run the `bootstrap` command and not the `createquestions` command, or vice versa? If the answer is "no", it would be nicer to not have to call this extra command from the docker startup, and rather include this utility in `bootstrap.py`.

**Review by BeritJanssen — APPROVED** [review]
Nice! The `createquestions` command looks as if it might be included in the `bootstrap` command, but perhaps there's arguments against it? Does retrieval of a single question from the question series work now, or is this left for later?
